### ISSUE TITLE
Roadmap to Stable-Baselines3 V1.0

### ISSUE THREAD

**Issue by araffin** [DLR-RM/stable-baselines3#1]
**This issue is meant to be updated as the list of changes is not exhaustive**

Dear all,

Stable-Baselines3 beta is now out :tada: ! This issue is meant to reference what is implemented and what is missing before a first major version.

As mentioned in the README, **before v1.0, breaking changes may occur**. I would like to encourage contributors (especially the maintainers) to make comments on how to improve the library before v1.0 (and maybe make some internal changes).

I will try to review the features mentioned in https://github.com/hill-a/stable-baselines/issues/576 (and https://github.com/hill-a/stable-baselines/issues/733)
and I will create issues soon to reference what is missing.

### What is implemented?

- [x] basic features (training/saving/loading/predict)
- [x] basic set of algorithms (A2C/PPO/SAC/TD3)
- [x] basic pre-processing (Box and Discrete observation/action spaces are handled)
- [x] callback support
- [x] complete benchmark for the continuous action case
- [x] basic rl zoo for training/evaluating plotting (https://github.com/DLR-RM/rl-baselines3-zoo)
- [x] consistent api
- [x] basic tests and most type hints
- [x] continuous integration (I'm in discussion with the organization admins for that)
- [x] handle more observation/action spaces #4 and #5 (thanks @rolandgvc)
- [x] tensorboard integration #9 (thanks @rolandgvc)
- [x] basic documentation and notebooks
- [x] automatic build of the documentation
- [x] Vanilla DQN #6  (thanks @Artemis-Skade)
- [x] Refactor off-policy critics to reduce code duplication #3 (see #78 )
- [x] DDPG #3
- [x] do a complete benchmark for the discrete case #49  (thanks @Miffyli !)
- [x] performance check for continuous actions #48 (even better than gSDE paper)
- [x]  get/set parameters for the base class (#138 )
- [x] clean up type-hints in docs #10 (cumbersome to read)
- [x] documenting the migration between SB and SB3 #11
- [x] finish typing some methods #175 
- [x] HER #8  (thanks @megan-klaiber)
- [x] finishing to update and clean the doc #166  (**help is wanted**)
- [x] finishing to update the notebooks and the tutorial #7  (I will do that, only HER notebook missing)

### What are the new features?

- [x] much cleaner base code (and no more warnings =D )
- [x] independent saving/loading/predict for policies
- [x] State-Dependent Exploration (SDE) for using RL directly on real robots (this is a unique feature, it was the starting point of SB3, I published a paper on that: https://arxiv.org/abs/2005.05719)
- [x] proper evaluation (using separate env) is included in the base class (using `EvalCallback`)
- [x] all environments are ``VecEnv``
- [x] better saving/loading (now can include the replay buffer and the optimizers)
- [x] any number of critics are allowed for SAC/TD3
- [x] custom actor/critic net arch for off-policy algos (#113 )
- [x] QR-DQN in [SB3-Contrib](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib)
- [x] Truncated Quantile Critics (TQC) (see #83 ) in [SB3-Contrib](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib)
- @Miffyli suggested a "contrib" repo for experimental features (it is [here](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib))

### What is missing?

- [x] syncing some files with Stable-Baselines to remain consistent (we may be good now, but need to be checked)
- [x] finish code-review of exisiting code #17 

### Checklist for v1.0 release

- [x] Update Readme
- [x]  Prepare blog post
- [x] Update doc: add links to the stable-baselines3 contrib
- [x]  Update docker image to use newer Ubuntu version
- [ ] Populate RL zoo

## What is next? (for V1.1+)

- basic dict/tuple support for observations (#243 )
- simple recurrent policies?
- DQN extensions (double, PER, IQN)
- Implement TRPO
- multi-worker training for all algorithms (#179 )
- n-step returns for off-policy algorithms #47  (@PartiallyTyped )
- SAC discrete #157  (need to be discussed, benefit vs DQN+extensions?)
- Energy Based Prioritisation? (@RyanRizzo96)
- implement `action_proba` in the base class?
- test the doc snippets #14  (**help is welcomed**)
- noisy networks (https://arxiv.org/abs/1706.10295) @PartiallyTyped ? exploration in parameter space?
- Munchausen Reinforcement Learning (MDQN) (probably in the contrib first, e.g. https://github.com/pfnet/pfrl/pull/74)

side note: should we change the default `start_method` to `fork`? (now that we don't have tf anymore)

**Comment by PartiallyTyped** [comment]
Maybe N-step returns for TD3 (and DDPG) and DQN (and friends) ? If it's implemented in the experience replay, then it is likely plug and play for TD3 and DQN, an implementation for SAC probably requires extra effort.

Perhaps at a later time, e.g. V1.1+ retrace, tree backup and q(lambda)?

**Comment by Miffyli** [comment]
@PartiallyTyped 

Yup, that would be v1.1 thing but indeed planned. Should probably go over original SB issues to gather all these suggestions at some point.

**Comment by PartiallyTyped** [comment]
Perhaps a discrete version of SAC for v1.1+?
https://arxiv.org/abs/1910.07207

**Comment by rolandgvc** [comment]
I will start working on the additional observation/action spaces this weekend 👍

**Comment by justinkterry** [comment]
Will the stable baselines 3 repo/package replace the existing stable baselines one, or will all this eventually be merged into the normal stable baselines package?

**Comment by Miffyli** [comment]
@justinkterry 

There are no plans to merge/combine the two repositories. Stable-baselines will continue to exist, and continue to receive bug-fixes and the like for some time before it is archived.

**Comment by justinkterry** [comment]
@Miffyli Thank you. Will it remain as "pip3 install stable-baselines," or become something like "pip3 install stable-baselines3"?

**Comment by Miffyli** [comment]
@justinkterry 

You can already install sb3 with `pip3 install stable-baseline3`. The original repo will stay as `pip3 install stable-baselines`.

**Comment by AdamGleave** [comment]
> * [ ]  tensorboard integration #9 (may be fast if we use the logger for that, maybe for 1.1+?), it should remain optional as it requires tensorflow :/ (**help is welcomed**)

I think if we log with e.g. [tensorboardX](https://github.com/lanpa/tensorboardX) there would be no need to have a TensorFlow dependency. The TensorBoard UI itself I believe does not require TensorFlow: it's [requirements.txt](https://github.com/tensorflow/tensorboard/blob/9fe0cbbe0e884c79848a90142bbaac7f726e646f/tensorboard/pip_package/requirements.txt) doesn't list it and in a fresh virtual environment after I `pip install tensorboard` I get:

```bash
$ pip list | grep tensorboard
tensorboard            2.2.1
tensorboard-plugin-wit 1.6.0.post3
$ pip list | grep tensorflow
$
```

That said, we should probably still keep this dependency optional.

**Comment by AdamGleave** [comment]
Minor point but I wonder if we should rename `BaseRLModel` to `BaseRLAlgorithm` and `BasePolicy` to `BaseModel`, given that `BaseRLPolicy` is more than just a policy?

**Comment by araffin** [comment]
>Minor point but I wonder if we should rename BaseRLModel to BaseRLAlgorithm and BasePolicy to BaseModel, given that BasePolicy is more than just a policy?

Good point, `BaseModel` and `BaseRLAlgorithm ` are definitely better names ;)

**Comment by jdily** [comment]
for visualization, probably using something like weights & biases (https://www.wandb.com/) is an option?
So that no need for tensorflow dependency.

**Comment by araffin** [comment]
>for visualization, probably using something like weights & biases (https://www.wandb.com/) is an option?

correct me if I'm wrong but W&B does not work offline, no? This is really important as you don't want your results to be published when you do private work.

This could be also implemented either as a callback (cf doc) or a new output for the logger. But sounds more like a "contrib" module to me.

**Comment by PartiallyTyped** [comment]
Perhaps an official shorthand for `stable-baselines` and `stable-baselines3` e.g. `sb` and `sb3`?

**Comment by ManifoldFR** [comment]
Is it necessary to continue to provide the interface for vectorized environments inside of this codebase?
They were contributed upstream back to gym [in this PR](https://github.com/openai/gym/pull/1513). After that PR was merged, packages such as PyBullet (`pybullet_envs`) started providing vectorized variants of their own environments using the interface from gym which should be the same as the one here (for now)

**Comment by Miffyli** [comment]
@ManifoldFR Somehow that has eluded my attention. Looks like a good suggestion! Less repeated code is better, as long it fits in stable-baselines functions too.

@araffin thoughts (you have most experience doing the eval/wrap functions)?

**Comment by ManifoldFR** [comment]
I happened onto it by chance because it's not documented anywhere inside of gym's docs, the openai people ported it from their own baselines repo with barely any notification of the change to end users.

**Comment by araffin** [comment]
So, I was aware of this (wrote some comments at that time https://github.com/openai/gym/pull/1513/files#r293899941) but I would argue against for different reasons:
- we rely on some specific features (set_attr, get_attr)
- openai version is undocumented and we don't know if they gonna break that feature (which is central in SB3) in a future release (I don't want to write new monkey patch like https://github.com/hill-a/stable-baselines/commit/678f803698de1a3735ded16f4a0c8d47e329465f)
- we can directly tweak that feature to fit our needs (and don't wait for a review and release by OpenAI)

So, in short, I would be in favor only if OpenAI way of maintaining Gym was more reliable.

**Comment by ManifoldFR** [comment]
I see, it figures that OpenAI changing things unilaterally without
documentation would be a problem.

I guess ensuring stable-baselines3's code doesn't break when running
vectorized envs derived from Gym's implementation would be safer and easier
instead

**Comment by PartiallyTyped** [comment]
How about vectorized/stacked action noise for TD3/SAC? I am referring to https://github.com/hill-a/stable-baselines/issues/805 . This will be a useful stepping stone for multiprocessing in TD3/SAC.

**Comment by araffin** [comment]
>How about vectorized/stacked action noise for TD3/SAC?

sounds good. If you have test included, you can do the PR.
It won't be used for a while though... a bit like what we did for dict support inside `VecEnv` (was not used until HER was re-implemented).

**Comment by PartiallyTyped** [comment]
For v1.1+: [Network Randomization](https://openreview.net/forum?id=HJgcvJBFvB) appears to be rather useful and simple enough to implement.  It is something that I will be implementing for a project I am working on so if you believe it will be useful to have I can open an issue specifically for it and discuss it there.

**Comment by PartiallyTyped** [comment]
For v1.1+
[Network Randomization](https://openreview.net/forum?id=HJgcvJBFvB) appears to be useful and simple enough to implement. I will implement it for a project I am working on (Obstacle Tower) so if you believe it will be useful to have, I can open an issue specifically for it and discuss it there.

**Comment by araffin** [comment]
>Network Randomization appears to be useful and simple enough to implement.

I would rather favor what we did for the pre-training (issue #27 ), create an example code/notebook and link to it in the documentation.

**Comment by PartiallyTyped** [comment]
> I would rather favor what we did for the pre-training (issue #27 ), create an example code/notebook and link to it in the documentation.

Sounds good to me. Perhaps this kind of ideas/projects could be moved to a contrib/example repository like Pytorch does?

**Comment by PartiallyTyped** [comment]
What do you think about Noisy Linear layers for v1.0?

**Comment by araffin** [comment]
>Noisy Linear layers

What is that?

**Comment by PartiallyTyped** [comment]
It's from [this](https://arxiv.org/abs/1706.10295) paper. The tl:dr is that they are linear layers that have parameter noise. This results in better exploration over e-greedy/ entropy bonus. The layers learn a mu and a logstd, then on the forward pass, they sample from the distribution.

**Comment by araffin** [comment]
>The tl:dr is that they are linear layers that have parameter noise. This results in better exploration over e-greedy/ entropy bonus. The layers learn a mu and a logstd, then on the forward pass, they sample from the distribution.

I considered that more as an extension to DQN (so v1.1+) but yes, why not (I added it to the roadmap). V1.0 is mainly meant to be the basic algorithms, fully tested and without extension.

In fact, from what I remember, this is very close to exploration in parameter space.
For continuous actions, gSDE is already implemented (see https://arxiv.org/abs/2005.05719), however, such exploration is missing for discrete actions.

**Comment by PartiallyTyped** [comment]
I agree on 1.1, it will be much easier to incorporate once the code is stable. I have it in internal code for PPO and I will be happy to make a PR when the time comes.

**Comment by AlessandroZavoli** [comment]
Do you think that algorithms like IMPALA or APEX are worth implementing?

**Comment by Miffyli** [comment]
Those _could_ be planned for v1.2+ and onwards, [as discussed here](https://github.com/hill-a/stable-baselines/issues/885). The thing is that the structure does not really fit well with the asynch nature of those algorithms.

**Comment by AlessandroZavoli** [comment]
multi-agent could be an very interesting step forward. 
I don't know if it is worth to develop a brand-new code if the structure may not fit this paradigm (the multi-agent, not one specific algorithm or another) that could be appealing for the community, given the fact that there are also few competitors.

**Comment by justinkterry** [comment]
Hey, so I've been managing the development of a library called PettingZoo. It's basically a multi-agent version of Gym, with a very Gym like API. It's involved people from UMD, Technical University Berlin, MILA, UT Austin, Google Research, and DeepMind. We even made changes to the ALE to allow playing multi-player Atari games with RL for the first time.

https://github.com/PettingZoo-Team/PettingZoo

Right now it's believed to mostly be in fully working order and is soft released, and we're aiming for a full release in the next month or so. This is something worth consideration in the planning of multi-agent support in stable baselines.

Additionally, me and several of the people at UMD have done a bunch with multi-agent RL, and have been wanting to add support for it to SB3 once it was more production ready. If there's interest in doing official multi-agent support with SB3, we should open a separate thread for the discussion of what that should look like.

**Comment by araffin** [comment]
> Hey, so I've been managing the development of a library called PettingZoo since January. It's basically a multi-agent version of Gym, with a very Gym like API. It's involved people from UMD, Technical University Berlin, MILA, UT Austin, Google Research, and DeepMind. We even made changes to the ALE to allow playing multi-player Atari games with RL for the first time.
> 
> https://github.com/PettingZoo-Team/PettingZoo
> 
> Right now it's believed to mostly be in fully working order and is soft released, and we're aiming for a full release in the next month or so. This is something worth consideration in the planning of multi-agent support in stable baselines.
> 
> Additionally, me and several of the people at UMD have done a bunch with multi-agent RL, and have been wanting to add support for it to SB3 once it was more production ready. If there's interest in doing official multi-agent support with SB3, we should open a separate thread for the discussion of what that should look like.

@justinkterry what you are doing is almost spam... You have been posting your project link already in 3 different issues.
We appreciate the effort of making a MA version of SB3 but this will be delayed to a separate issue.
In order to keep this thread clean and focus on the 1.0 roadmap, i will delete your and my comment, feel free to open a new issue for discussion about MA support.

**Comment by justinkterry** [comment]
There have been three different isolated discussions of MA support for stable baselines afoot. I made a brief mention of a relevant project in each because I wanted it to be considered in whatever highly decentralized discussions that are going on about this topic.

I apologize if that's annoying, but it hardly seems to rise to the level of spam.

**Comment by araffin** [comment]
I move the discussion of MA/Distributed agents support to https://github.com/DLR-RM/stable-baselines3/issues/69
I deleted some comments to re-focus on 1.0 roadmap.

**Comment by araffin** [comment]
I have some good news: performances of all algorithms are matching the ones from Stable-Baselines (SB2) both in discrete and continuous environments :tada: !
See #48 and #49 (with some nice detective work from @Miffyli in #110 )
I think I will add a note in the README to show that SB3 is now trustworthy ;)

I could even reproduce results (in fact have even better ones after the bug fixes) from the [gSDE](https://paperswithcode.com/paper/generalized-state-dependent-exploration-for) paper

**Comment by AlessandroZavoli** [comment]
Very good! What were the problems/bugs?

**Comment by ManifoldFR** [comment]
One interesting feature would be to supply different inputs to the policy and value networks, when these are separate. This is the principle behind [Asymmetric Actor Critic](https://arxiv.org/abs/1710.06542), which is used at OpenAI to make the critic omniscient and get a somewhat better advantage estimate during policy optimization

**Comment by araffin** [comment]
>One interesting feature would be to supply different inputs to the policy and value networks, when these are separate. This is the principle behind Asymmetric Actor Critic, which is used at OpenAI to make the critic omniscient and get a somewhat better advantage estimate during policy optimization

The scope of this method is unfortunately very narrow (only applies when you want to do sim2real and have access to the simulator) compared to the amount of changes needed.

**Comment by cosmir17** [comment]
@PartiallyTyped I found the original example for the SAC Discrete Implementation plan.

Can the following paper be considered?
https://arxiv.org/abs/1910.07207

The author also publicised his code.
https://github.com/p-christ/Deep-Reinforcement-Learning-Algorithms-with-PyTorch/blob/master/agents/actor_critic_agents/SAC_Discrete.py

Hope these help,
Sean

**Comment by araffin** [comment]
>I found the original example for the SAC Discrete Implementation plan.
>Can the following paper be considered?
>https://arxiv.org/abs/1910.07207


We already have an issue for that #157

**Comment by cosmir17** [comment]
@araffin I wasn't asking if we can implement it. Given that it has been decided that stable-baselines3 is going to do it and PartialTyped volunteered, I was giving him or the team a resource.

**Comment by araffin** [comment]
First release candidate is out: https://github.com/DLR-RM/stable-baselines3/releases/tag/v1.0rc0
100+ trained rl models will be published soon: https://github.com/DLR-RM/rl-baselines3-zoo/pull/69

**Comment by ziegenbalg** [comment]
Excellent work guys! So is there a full example of A2C (or other) games using a Dict observation space? I tried modifying the test_dict_env.py but end up with a

`Traceback (most recent call last):
  File "/home/eziegenbalg/Documents/goldengoose_research/stable-baselines3/tests/test_dict_env.py", line 125, in <module>
    test_dict_spaces(A2C, True)
  File "/home/eziegenbalg/Documents/goldengoose_research/stable-baselines3/tests/test_dict_env.py", line 119, in test_dict_spaces
    model = model_class("MlpPolicy", env, gamma=0.5, seed=1, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/stable_baselines3/a2c/a2c.py", line 80, in __init__
    super(A2C, self).__init__(
  File "/usr/local/lib/python3.9/site-packages/stable_baselines3/common/on_policy_algorithm.py", line 76, in __init__
    super(OnPolicyAlgorithm, self).__init__(
  File "/usr/local/lib/python3.9/site-packages/stable_baselines3/common/base_class.py", line 156, in __init__
    env = self._wrap_env(env, self.verbose, monitor_wrapper)
  File "/usr/local/lib/python3.9/site-packages/stable_baselines3/common/base_class.py", line 209, in _wrap_env
    env = ObsDictWrapper(env)
  File "/usr/local/lib/python3.9/site-packages/stable_baselines3/common/vec_env/obs_dict_wrapper.py", line 45, in __init__
    dimensions = [venv.observation_space.spaces["observation"].n, venv.observation_space.spaces["desired_goal"].n]
KeyError: 'observation'`

Error.

**Comment by Miffyli** [comment]
If you want a more practical example, [see this comment that has code using ViZDoom in it](https://github.com/DLR-RM/stable-baselines3/pull/243#issuecomment-766459883). Other than that there is the documentation that has examples on using and modifying things.

**Comment by ziegenbalg** [comment]
@Miffyli, thank you for your quick reply. Am I understanding this correctly.. You are wrapping your Doom Dict Env in a DummyVecEnv? Why not just pass the Doom Env straight to PPO()?

**Comment by Miffyli** [comment]
That is used to gather rollout samples from multiple environments at the same time (leads to stabler training with PPO). SB3 also wraps all environments, even if you only have a single one, into a VecEnv under the hood.

**Comment by ziegenbalg** [comment]
Ok first off... are you a bot? So there was a specific reason you wrapped it in DummyVecEnv and not DummyDictEnv? I'll take a look at the difference between those DummyEnv now. I will try out the effects of wrapping with and without DummyVecEnv/DummyDictEnv with A2C and report back here for future truth seekers.

**Comment by Miffyli** [comment]
> Ok first off... are you a bot?

Nah, just happen to be around when you are sending messages :)

> So there was a specific reason you wrapped it in DummyVecEnv and not DummyDictEnv?

Yes, I see now the issue. These two things are very different: DummyDictEnv is just a testing environment used to test the support for dict-observations. DummyVecEnv is a VecEnv implementation that does not really parallelize anything. There is no connection between the two. You should _not_ use DummyDictEnv (it is only meant for testing), you should use DummyVecEnv (or SubProcVecEnv) when you give environments to SB3 code.
